File: src/main/java/org/moinex/ui/dialog/goal/BaseGoalManagement.java
/*
 * Filename: BaseGoalManagement.java
 * Created on: March  9, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.goal;

import java.util.List;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.GoalService;
import org.moinex.service.I18nService;
import org.moinex.service.WalletService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.springframework.beans.factory.annotation.Autowired;

/**
 * Base class to implement the common behavior of the Add and Edit Goal
 */
@NoArgsConstructor
public abstract class BaseGoalManagement {
    @FXML protected ComboBox<Wallet> masterWalletComboBox;

    @FXML protected TextField nameField;

    @FXML protected TextField balanceField;

    @FXML protected TextField targetBalanceField;

    @FXML protected DatePicker targetDatePicker;

    @FXML protected TextArea motivationTextArea;

    protected List<Wallet> masterWallets;

    protected WalletService walletService;

    protected GoalService goalService;

    protected I18nService i18nService;

    /**
     * Constructor
     *
     * @param goalService GoalService
     * @param walletService WalletService
     * @note This constructor is used for dependency injection
     */
    @Autowired
    protected BaseGoalManagement(GoalService goalService, WalletService walletService) {
        this.goalService = goalService;
        this.walletService = walletService;
    }

    protected void setI18nService(I18nService i18nService) {
        this.i18nService = i18nService;
    }

    @FXML
    protected void initialize() {
        UIUtils.setDatePickerFormat(targetDatePicker);

        loadWalletsFromDatabase();

        configureComboBoxes();

        populateComboBoxes();

        // Ensure that the balance fields only accept monetary values
        balanceField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.MONETARY_VALUE_REGEX)) {
                                balanceField.setText(oldValue);
                            }
                        });

        targetBalanceField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.MONETARY_VALUE_REGEX)) {
                                targetBalanceField.setText(oldValue);
                            }
                        });
    }

    @FXML
    protected abstract void handleSave();

    @FXML
    protected void handleCancel() {
        Stage stage = (Stage) nameField.getScene().getWindow();
        stage.close();
    }

    private void populateComboBoxes() {
        masterWalletComboBox.getItems().add(null);
        masterWalletComboBox.getItems().addAll(masterWallets);

        masterWalletComboBox.setCellFactory(
                lv ->
                        new ListCell<>() {
                            @Override
                            protected void updateItem(Wallet item, boolean empty) {
                                super.updateItem(item, empty);
                                if (empty || item == null) {
                                    setText(""); // Display empty text when no wallet is selected
                                } else {
                                    setText(item.getName());
                                }
                            }
                        });

        masterWalletComboBox.setButtonCell(
                new ListCell<>() {
                    @Override
                    protected void updateItem(Wallet item, boolean empty) {
                        super.updateItem(item, empty);
                        if (empty || item == null) {
                            setText(""); // Display empty text when no wallet is selected
                        } else {
                            setText(item.getName());
                        }
                    }
                });
    }

    private void configureComboBoxes() {
        UIUtils.configureComboBox(masterWalletComboBox, Wallet::getName);
    }

    private void loadWalletsFromDatabase() {
        masterWallets =
                walletService.getAllNonArchivedWalletsOrderedByName().stream()
                        .filter(Wallet::isMaster)
                        .toList();
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/goal/base/BaseGoalManagement.kt
/*
 * Filename: BaseGoalManagement.kt (original filename: BaseGoalManagement.java)
 * Created on: March  9, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.goal.base

import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.DatePicker
import javafx.scene.control.TextArea
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constants.Constants
import org.moinex.common.util.UIUtils
import org.moinex.model.wallettransaction.Wallet
import org.moinex.service.PreferencesService
import org.moinex.service.goal.GoalService
import org.moinex.service.wallet.WalletService

abstract class BaseGoalManagement(
    protected val goalService: GoalService,
    protected val walletService: WalletService,
    protected val preferencesService: PreferencesService,
) {
    @FXML
    protected lateinit var masterWalletComboBox: ComboBox<Wallet>

    @FXML
    protected lateinit var nameField: TextField

    @FXML
    protected lateinit var balanceField: TextField

    @FXML
    protected lateinit var targetBalanceField: TextField

    @FXML
    protected lateinit var targetDatePicker: DatePicker

    @FXML
    protected lateinit var motivationTextArea: TextArea

    protected var masterWallets: List<Wallet> = emptyList()

    @FXML
    protected open fun initialize() {
        UIUtils.setDatePickerFormat(targetDatePicker)

        loadWalletsFromDatabase()
        configureComboBoxes()
        populateComboBoxes()

        balanceField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.MONETARY_VALUE_REGEX))) {
                balanceField.text = oldValue
            }
        }

        targetBalanceField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.MONETARY_VALUE_REGEX))) {
                targetBalanceField.text = oldValue
            }
        }
    }

    @FXML
    protected abstract fun handleSave()

    @FXML
    protected open fun handleCancel() {
        (nameField.scene.window as Stage).close()
    }

    private fun populateComboBoxes() {
        masterWalletComboBox.items.add(null)
        masterWalletComboBox.items.addAll(masterWallets)

        masterWalletComboBox.setCellFactory {
            UIUtils.createListCell { wallet -> wallet?.name ?: "" }
        }

        masterWalletComboBox.buttonCell = UIUtils.createListCell { wallet -> wallet?.name ?: "" }
    }

    private fun configureComboBoxes() {
        UIUtils.configureComboBox(masterWalletComboBox, Wallet::name)
    }

    private fun loadWalletsFromDatabase() {
        masterWallets =
            walletService
                .getAllNonArchivedWalletsOrderedByName()
                .filter { it.isMaster() }
    }
}
